### PR TITLE
Make the configuration file optional

### LINKED ISSUE #5899: Suppress configuration file loading
Is it possible to suppress the loading of configuration files?

If not, what would be the best way to implement such functionality?
Would to PR...

### LINKED ISSUE #6511: XDG_CONFIG_HOME failing silently caused confusion
I ran into a group who would have benefited from #5899. However, since that is not yet possible, they tried to set two different directories for different versions of Astropy to run in parallel on the same file system on the same cluster. Otherwise, they run into a problem of `astropy.cfg` getting corrupted by the parallel attempts to write it.

Despite instructions at http://astropy.readthedocs.io/en/latest/config/index.html that says, "... if you’ve set the environment variable `XDG_CONFIG_HOME` and the `$XDG_CONFIG_HOME/astropy` directory exists, it will instead be there," they set `XDG_CONFIG_HOME` that included the "astropy" sub-directory. As a result, the check for `$XDG_CONFIG_HOME/astropy` failed (because it was really looking for `$XDG_CONFIG_HOME/astropy/astropy`) and fell back silently to the default `~/.astropy/config`, which caused a lot of confusion.

Suggestion: If `XDG_CONFIG_HOME` is defined but the desired directory is not found, raise an exception (too much?) or at least emit a warning.

### LINKED ISSUE #7290: Concurrent import of astropy occasionally fails with likely race condition on config files.
When trying to import astropy from an MPI application ( with 128 ranks), the application crashes during first import randomly with a file system error.

It appears that some astropy instances are trying to modify a configuration file while some others are trying to read it. 

```
    from astropy.cosmology import Planck13, Planck15, WMAP5, WMAP7, WMAP9
  File "/dev/shm/local/lib/python3.6/site-packages/astropy/__init__.py", line 286, in <module>
    log = _init_log()
  File "/dev/shm/local/lib/python3.6/site-packages/astropy/logger.py", line 97, in _init_log
    log._set_defaults()
  File "/dev/shm/local/lib/python3.6/site-packages/astropy/logger.py", line 473, in _set_defaults
    self.setLevel(conf.log_level)
  File "/dev/shm/local/lib/python3.6/site-packages/astropy/config/configuration.py", line 273, in __get__
    return self()
  File "/dev/shm/local/lib/python3.6/site-packages/astropy/config/configuration.py", line 396, in __call__
    sec = get_config(self.module)
  File "/dev/shm/local/lib/python3.6/site-packages/astropy/config/configuration.py", line 530, in get_config
    cobj = configobj.ConfigObj(cfgfn, interpolation=False)
  File "/dev/shm/local/lib/python3.6/site-packages/astropy/extern/configobj/configobj.py", line 1228, in __init__
    self._load(infile, configspec)
  File "/dev/shm/local/lib/python3.6/site-packages/astropy/extern/configobj/configobj.py", line 1317, in _load
    raise error
astropy.extern.configobj.configobj.ParseError: Invalid line ('\x00\x00\x00\x00\x00\x00\x00\x00\x00
....
```

### PR DESCRIPTION
**Pull request by saimn**
Ref discussion in #10738 and #10090 : stop writing/updating the configuration file by default when importing the package, and do nothing when reading if the file is absent.
I also replaced `update_default_config` with a new function `create_config_file` which can be used to create or update the default file. And removed some old stuff for versions < 0.4. And removed the default config file which is present in the package since it can be generated, this will avoid having to update it in the release procedure.

cc @astrofrog  @pllim @embray 

EDIT:
Fix #5899
Close #7290
Close #6511
Close #10090

### CODE REVIEW

File: astropy/config/configuration.py
Comment (astrofrog) [inline astropy/config/configuration.py:681]: We could make it even simpler and just error if the file already exists unless force is given, without checking content? Would simplify the code further...

**Review by astrofrog — COMMENTED** [review]
Just a quick comment below and also wanted to say that we should make sure we aren't breaking anything for downstream packages (the change in behaviour is fine, I mean no removal of functions etc that might break downstream)

**Comment by pllim** [comment]
@saimn , since title says WIP, I converted this PR to draft. FYI.

File: astropy/config/configuration.py
Comment (pllim) [reply astropy/config/configuration.py:681]: In my experience though, if user really comes across this exception, we will have a feature request to ask for option to overwrite... 🤷

File: astropy/config/configuration.py
Comment (pllim) [inline astropy/config/configuration.py:689]: More consistent with our existing API over at unified IO?
```suggestion
    overwrite : bool
```

**Review by pllim — COMMENTED** [review]
I do like the amount of simplification this introduces, but given the behavior change, this should be in 4.3 or later, at least. I have set the milestone to 5.0 for now, but it is open for discussion.

User facing doc also needs to be written to fully explain what to expect with this change. Perhaps even a section for downstream package maintainers.

This patch should be tested with coordinated packages and a few affiliated ones (I would want to test mine at least) before being merged. The consequences of setting config alias, adding new config, removing old config, or changing existing config data type should also be considered.

**Comment by saimn** [comment]
Thanks @pllim, I could not find how to make it a draft PR...

**Comment by saimn** [comment]
> @astrofrog - Just a quick comment below and also wanted to say that we should make sure we aren't breaking anything for downstream packages (the change in behaviour is fine, I mean no removal of functions etc that might break downstream)

Yep, I released after that astroquery may use `update_default_config` (and it does: https://github.com/astropy/astroquery/blob/master/astroquery/_astropy_init.py#L43). astroquery is probably the only package using the configuration system ?
So we could keep `update_default_config` longer, or make it do nothing (keeping it as an empty function would stop updating the config file as for astropy core) ?

**Comment by pllim** [comment]
> astroquery is probably the only package using the configuration system

Actually, a few of my own packages relies on it. For example, `synphot` and `stsynphot`.

> draft PR

Click the small arrow on the giant green button that says "create PR" and there will be a second option to create draft PR. 😉

File: astropy/config/configuration.py
Comment (saimn) [reply astropy/config/configuration.py:681]: Yes, I'm happy to simplify more.

Comment (saimn) [reply astropy/config/configuration.py:681]: Thinking more about this, the problem is that the direct string comparison will fail if a configuration item has been added. So `is_unedited_config_file` is a nice a simple way to check if the config file has not been customized, whatever the version was used to generate it.

File: astropy/config/configuration.py
Comment (saimn) [reply astropy/config/configuration.py]: Done.

File: astropy/config/configuration.py
Comment (saimn) [reply astropy/config/configuration.py]: I restored `update_default_config` since other packages use it, and it relies on `is_unedited_config_file`.

**Comment by saimn** [comment]
> I do like the amount of simplification this introduces, but given the behavior change, this should be in 4.3 or later, at least. I have set the milestone to 5.0 for now, but it is open for discussion.

5.0 seems a bit far way, and I don't think this change would be a big disruption ;)

> User facing doc also needs to be written to fully explain what to expect with this change. Perhaps even a section for downstream package maintainers.

Sure!

> This patch should be tested with coordinated packages and a few affiliated ones (I would want to test mine at least) before being merged. The consequences of setting config alias, adding new config, removing old config, or changing existing config data type should also be considered.

Not sure what you mean with config alias and data type ?
I think the main thing to test is that the `generate_config`/`create_config_file` functions work well for other packages. I'm trying with astroquery, and there are some issues ;) (one of the subpackages is listed 3 times in the generated config file), so I will try to fix that.

**Comment by astrofrog** [comment]
What about milestoning for 4.3 and then merging as soon as possible after the 4.2 branching which will give ample time for testing with downstream packages?

**Comment by saimn** [comment]
Ref #10893 for a few fixes to get `generate_config` works correctly with astroquery.

File: astropy/config/configuration.py
Comment (embray) [inline astropy/config/configuration.py:822]: Instead of comparing the config files byte-for-byte, why not just parse the existing config file and compare it to the default configuration?

This might require some refactoring of `generate_config` to have it generate the actual config data structure first, then a separate method for writing it to a file; probably how it should have been done in the first place.

Unless we really care about ensuring that it is byte-for-byte identical.

**Review by embray — COMMENTED** [review]
:+1: aside from this one question.

File: astropy/config/configuration.py
Comment (saimn) [reply astropy/config/configuration.py:822]: Yes, that's the way it was done. But just after `is_unedited_config_file` loads the config file and checks if any keys are defined. So, since now this will not be run at import, maybe we can just keep the second check?

**Comment by saimn** [comment]
I added a test and updated the documentation, so I think this is now ready. Please review again @pllim @embray @astrofrog

File: CHANGES.rst
Comment (embray) [inline CHANGES.rst:338]: "no more" -> "no longer" x2

File: astropy/config/configuration.py
Comment (embray) [inline astropy/config/configuration.py:46]: Removing this might break a tiny handful of existing scripts that deliberately import `ConfigurationMissingWarning` in order to suppress it.  I'm not sure if there is a great (existing) way we could create a deprecation warning for it in this case though.

Comment (embray) [reply astropy/config/configuration.py:46]: Maybe this could just be noted in the changelog.

File: astropy/config/configuration.py
Comment (embray) [inline astropy/config/configuration.py:844]: Why does it use latin-1?

File: docs/config/index.rst
Comment (embray) [inline docs/config/index.rst:19]: "no more" -> "no longer"

**Review by embray — DISMISSED** [review]

File: astropy/config/configuration.py
Comment (saimn) [reply astropy/config/configuration.py:46]: Right, we can keep it for now. Btw I should probably add the deprecation warning for `update_default_config` ?

File: astropy/config/configuration.py
Comment (saimn) [reply astropy/config/configuration.py:844]: No idea. I guess this comes from the old times with Python 2 etc. Since the plan is to remove this function in the future, I guess we can keep it as it is for now.

File: CHANGES.rst
Comment (saimn) [reply CHANGES.rst]: Done, same below, thanks.

**Comment by pllim** [comment]
The failures make no sense to me and it is unsettling because I see them in `master` too. I'll hold off review until we can fix the CI.

**Comment by saimn** [comment]
@pllim - See #11305.

**Comment by pllim** [comment]
@saimn , please rebase. 😉

**Comment by saimn** [comment]
I'm not sure if the deprecation warning for `update_default_config` is a good idea... A lot of packages use this function as it's used by the template (https://github.com/search?l=Python&q=update_default_config&type=code) so we should first removed it from the template and give time to projects to adopt those changes.

**Comment by pllim** [comment]
Thanks, @saimn ! @embray , since you requested changes, do you want to re-review?

**Comment by pllim** [comment]
There is a conflict now... From the Coordination Meeting 2021, there is desire to revisit this and get it merged.

**Comment by taldcroft** [comment]
To elaborate on what @pllim said, there was specific feedback that writing the config file is a problem in read-only environments like certain HPC platforms. So 👍 from me on the concept (but I plead ignorance on code details).

**Comment by saimn** [comment]
I removed the deprecation warning, please review again.

**Review by pllim — COMMENTED** [review]
Some comments/questions below. Also:

* Will https://docs.astropy.org/en/latest/config/astropy_config.html#astropy-config-file still be generated?

After this PR is merged, need to open issue to follow up with deprecation and removal of `update_default_config` and `ConfigurationMissingWarning` after communicating the plan to affected downstream packages

File: docs/config/index.rst
Comment (saimn) [reply docs/config/index.rst:20]: Probably not, maybe for historical purpose ?

Comment (pllim) [reply docs/config/index.rst:20]: I vote that we just remove mention of 0.4. Pretty sure no one uses it now.

File: docs/config/index.rst
Comment (saimn) [reply docs/config/index.rst:227]: Nothing bad happens because the default values will be used for configuration items that are not in the file. Same as now since config items are commented by default. So the only thing that will happen is that users may have an old configuration file from an old version, and thus they may miss some configuration items that will be added in or after 4.3. But we don't add config item so often, and the documentation is there to explain how to update this file if needed.

**Comment by saimn** [comment]
image
:astonished:

**Comment by saimn** [comment]
Oh, that's the master→default rename ...

File: docs/config/index.rst
Comment (pllim) [reply docs/config/index.rst:227]: Well, I guess it is not clear to me whether the doc is meant for end-user or downstream package maintainers, but maybe it is just me.

File: docs/config/index.rst
Comment (pllim) [reply docs/config/index.rst:20]: I need to merge this pronto to fix CI, so I guess this note entry survives another day.

**Review by pllim — APPROVED** [review]
Let's do it! ✔️

File: docs/config/index.rst
Comment (saimn) [reply docs/config/index.rst:227]: Right, but for the configuration system I'm not sure if downstream packages have other documentation, I guess astropy documentation is the reference one.

**Pull request merged**
